test/unit/metadata_test.py

```diff
@@ -232,6 +236,108 @@ def test_frictionless_data_package_resources_populated():
+    assert partition_offsets[partition_key](period, offset) == expected
+
+
+@pytest.mark.parametrize("source_id", sorted(r for r in SOURCES))
+def test_source_availability(source_id):
+    """Check that all sources have a reasonable temporal availability.
+
+    Sources with only non-temporal partitions will evaluate to None; all others
+    should show after 1990.
+
+    We check this because if you get the data types wrong in pd.Timestamp, it
+    spits out 1970 instead of the proper year.
+    """
+    src = DataSource.from_id(source_id)
+    availability = ResourceDescriptionBuilder.offset_source_availability(src, 0)
+    # Checking the lexical ordering of strings using > is a bit brittle, but has
+    # the bonus of handling years, year quarters, half years, and year months.
+    # If this breaks it probably means we're running on a machine that orders
+    # strings by weird criteria -- we can revisit this at that time.
+    assert (availability is None) or (availability > "1990")
+
+
+@pytest.mark.parametrize(
+    "given_name,given_settings,given_rowcounts,given_source,expected",
+    [
+        (  # manually specified
+            "test_eia__entity_test",
+            {"availability_text": "manual", "sources": []},
+            [None],
+            [],
+            ResourceTrait(type="True", description="manual"),
+        ),
+        (  # using row counts
+            "test_eia__entity_test",
+            {"sources": []},
+            ["row counts"],
+            [],
+            ResourceTrait(type="True", description="row counts"),
+        ),
+        (  # using source: least recent among at least one temporal partition
+            "test_eia__entity_test",
+            {"sources": ["1", "x", "2"]},
+            [None],
+            ["1 source", None, "2 source"],
+            ResourceTrait(type="True", description="1 source"),
+        ),
+        (  # using source; none temporal
+            "test_eia__entity_test",
+            {"sources": ["x", "y"]},
+            [None],
+            [None, None],
+            ResourceTrait(type="False", description="Unknown"),
+        ),
+        (  # nobody knows
+            "test_eia__entity_test",
+            {"sources": []},
+            [None],
+            [],
+            ResourceTrait(type="False", description="Unknown"),
+        ),
+    ],
+)
+def test_resolve_resource_availability(
+    mocker, given_name, given_settings, given_rowcounts, given_source, expected
+):
+    """Verify fallback flow is correct for resolving most recently available data."""
+    mocker.patch(
+        "pudl.metadata.descriptions.ResourceDescriptionBuilder.compute_rowcounts_availability",
+        mocker.Mock(side_effect=given_rowcounts, __is_component=False),
+    )
+    mocker.patch(
+        "pudl.metadata.descriptions.ResourceDescriptionBuilder.offset_source_availability",
+        mocker.Mock(side_effect=given_source, __is_component=False),
+    )
```

Comment:
Don't actually read from row counts file or source metadata -- we're just exercising edge cases here